### ISSUE TITLE
SecurityException: class "com.microsoft.applicationinsights.web.javaee.RequestNameInterceptor_cs27u3_Impl"'s signer information does not match signer information of other classes in the same package - Weblogic

### ISSUE THREAD

**Issue by ijmac3** [microsoft/ApplicationInsights-Java#1179]
Our product is currently required to work on both Wildfly and Weblogic. Previously, we were only using `applicationinsights-core` and `applicationinsights-logging-log4j2` and this was working well. Recently, we also introduced the `applicationinsights-web-auto` library and it stopped working on Weblogic, complaining that `com.microsoft.applicationinsights.web.javaee.RequestNameInterceptor_cs27u3_Impl`'s signer information didn't match that of other classes in the same package. This issue does not occur with Wildfly.

### Expected behavior
Application starts on Weblogic without issue.

### Actual behavior
Throws a SecurityException and fails to start (see Logs for full stacktrace).

### System information
 - ApplicationInsights version 2.5.1 (also tried with 2.6.0-BETA)
 - Application Server type and version (if applicable): Weblogic 12.2.1.3.0 / Wildfly 15.0.1-FINAL
 - Java 8
 - OS type and version: Windows 10

 - Using spring-boot? - No

### Logs
```
<17-Dec-2019 09:37:57,165 o'clock GMT> <Warning> <EJB> <BEA-011157> <BEA-011157> <An exception occurred while preparing the EJB module swc.jar.

Unable to deploy EJB: swc.jar from swc.jar:

There are 1 nested errors:

java.lang.SecurityException: class "com.microsoft.applicationinsights.web.javaee.RequestNameInterceptor_cs27u3_Impl"'s signer information does not match signer information of other classes in the same package
	at java.lang.ClassLoader.checkCerts(ClassLoader.java:898)
	at java.lang.ClassLoader.preDefineClass(ClassLoader.java:668)
	at java.lang.ClassLoader.defineClass(ClassLoader.java:761)
	at java.security.SecureClassLoader.defineClass(SecureClassLoader.java:142)
	at weblogic.utils.classloaders.GenericClassLoader.defineClassInternal(GenericClassLoader.java:1113)
	at weblogic.utils.classloaders.GenericClassLoader.defineClass(GenericClassLoader.java:1046)
	at weblogic.utils.classloaders.GenericClassLoader.findLocalClass(GenericClassLoader.java:1038)
	at weblogic.utils.classloaders.GenericClassLoader.findClass(GenericClassLoader.java:990)
	at weblogic.utils.classloaders.GenericClassLoader.doFindClass(GenericClassLoader.java:611)
	at weblogic.utils.classloaders.GenericClassLoader.loadClass(GenericClassLoader.java:543)
	at weblogic.utils.classloaders.GenericClassLoader.loadClass(GenericClassLoader.java:496)
	at weblogic.utils.classloaders.GenericClassLoader.loadClass(GenericClassLoader.java:473)
	at java.lang.Class.forName0(Native Method)
	at java.lang.Class.forName(Class.java:348)
	at weblogic.utils.classloaders.GenericClassLoader.defineCodeGenClass(GenericClassLoader.java:1360)
	at weblogic.ejb.container.ejbc.bytecodegen.GeneratorFactory.generateClass(GeneratorFactory.java:100)
	at weblogic.ejb.container.ejbc.bytecodegen.GeneratorFactory.generate(GeneratorFactory.java:44)
	at weblogic.ejb.container.ejbc.EJBCompiler.generate(EJBCompiler.java:577)
	at weblogic.ejb.container.ejbc.EJBCompiler.doCompile(EJBCompiler.java:186)
	at weblogic.ejb.container.ejbc.EJBCompiler.compileEJB(EJBCompiler.java:336)
	at weblogic.ejb.container.ejbc.EJBCompiler.compileEJB(EJBCompiler.java:304)
	at weblogic.ejb.container.deployer.EJBDeployer.compileEjbs(EJBDeployer.java:424)
	at weblogic.ejb.container.deployer.EJBDeployer.compileIfNecessary(EJBDeployer.java:355)
	at weblogic.ejb.container.deployer.EJBDeployer.prepare(EJBDeployer.java:723)
	at weblogic.ejb.container.deployer.EJBModule.prepare(EJBModule.java:405)
	at weblogic.application.internal.ExtensibleModuleWrapper$PrepareStateChange.next(ExtensibleModuleWrapper.java:295)
	at weblogic.application.internal.ExtensibleModuleWrapper$PrepareStateChange.next(ExtensibleModuleWrapper.java:285)
	at weblogic.application.utils.StateMachineDriver.nextState(StateMachineDriver.java:45)
	at weblogic.application.internal.ExtensibleModuleWrapper.prepare(ExtensibleModuleWrapper.java:109)
	at weblogic.application.internal.flow.ModuleListenerInvoker.prepare(ModuleListenerInvoker.java:100)
	at weblogic.application.internal.flow.ModuleStateDriver$1.next(ModuleStateDriver.java:192)
	at weblogic.application.internal.flow.ModuleStateDriver$1.next(ModuleStateDriver.java:187)
	at weblogic.application.utils.StateMachineDriver$ParallelChange.run(StateMachineDriver.java:83)
	at weblogic.application.utils.StateMachineDriver.nextStateInParallel(StateMachineDriver.java:144)
	at weblogic.application.internal.flow.ModuleStateDriver.parallelPrepare(ModuleStateDriver.java:46)
	at weblogic.application.internal.flow.DeploymentCallbackFlow.prepare(DeploymentCallbackFlow.java:75)
	at weblogic.application.internal.flow.DeploymentCallbackFlow.prepare(DeploymentCallbackFlow.java:55)
	at weblogic.application.internal.BaseDeployment$1.next(BaseDeployment.java:729)
	at weblogic.application.utils.StateMachineDriver.nextState(StateMachineDriver.java:45)
	at weblogic.application.internal.BaseDeployment.prepare(BaseDeployment.java:241)
	at weblogic.application.internal.EarDeployment.prepare(EarDeployment.java:66)
	at weblogic.application.internal.DeploymentStateChecker.prepare(DeploymentStateChecker.java:158)
	at weblogic.deploy.internal.targetserver.AppContainerInvoker.prepare(AppContainerInvoker.java:65)
	at weblogic.deploy.internal.targetserver.AppDeployment.prepare(AppDeployment.java:166)
	at weblogic.management.deploy.internal.DeploymentAdapter$1.doPrepare(DeploymentAdapter.java:41)
	at weblogic.management.deploy.internal.DeploymentAdapter.prepare(DeploymentAdapter.java:193)
	at weblogic.management.deploy.internal.AppTransition$1.transitionApp(AppTransition.java:31)
	at weblogic.management.deploy.internal.ConfiguredDeployments$2.doItem(ConfiguredDeployments.java:741)
	at weblogic.management.deploy.internal.parallel.BucketInvoker$2.run(BucketInvoker.java:95)
	at weblogic.work.ContextWrap.run(ContextWrap.java:46)
	at weblogic.work.SelfTuningWorkManagerImpl$WorkAdapterImpl.run(SelfTuningWorkManagerImpl.java:670)
	at weblogic.invocation.ComponentInvocationContextManager._runAs(ComponentInvocationContextManager.java:352)
	at weblogic.invocation.ComponentInvocationContextManager.runAs(ComponentInvocationContextManager.java:337)
	at weblogic.work.LivePartitionUtility.doRunWorkUnderContext(LivePartitionUtility.java:57)
	at weblogic.work.PartitionUtility.runWorkUnderContext(PartitionUtility.java:41)
	at weblogic.work.SelfTuningWorkManagerImpl.runWorkUnderContext(SelfTuningWorkManagerImpl.java:644)
	at weblogic.work.ExecuteThread.execute(ExecuteThread.java:415)
	at weblogic.work.ExecuteThread.run(ExecuteThread.java:355)

	at weblogic.ejb.container.deployer.EJBDeployer.compileEjbs(EJBDeployer.java:426)
	at weblogic.ejb.container.deployer.EJBDeployer.compileIfNecessary(EJBDeployer.java:355)
	at weblogic.ejb.container.deployer.EJBDeployer.prepare(EJBDeployer.java:723)
	at weblogic.ejb.container.deployer.EJBModule.prepare(EJBModule.java:405)
	at weblogic.application.internal.ExtensibleModuleWrapper$PrepareStateChange.next(ExtensibleModuleWrapper.java:295)
	at weblogic.application.internal.ExtensibleModuleWrapper$PrepareStateChange.next(ExtensibleModuleWrapper.java:285)
	at weblogic.application.utils.StateMachineDriver.nextState(StateMachineDriver.java:45)
	at weblogic.application.internal.ExtensibleModuleWrapper.prepare(ExtensibleModuleWrapper.java:109)
	at weblogic.application.internal.flow.ModuleListenerInvoker.prepare(ModuleListenerInvoker.java:100)
	at weblogic.application.internal.flow.ModuleStateDriver$1.next(ModuleStateDriver.java:192)
	at weblogic.application.internal.flow.ModuleStateDriver$1.next(ModuleStateDriver.java:187)
	at weblogic.application.utils.StateMachineDriver$ParallelChange.run(StateMachineDriver.java:83)
	at weblogic.application.utils.StateMachineDriver.nextStateInParallel(StateMachineDriver.java:144)
	at weblogic.application.internal.flow.ModuleStateDriver.parallelPrepare(ModuleStateDriver.java:46)
	at weblogic.application.internal.flow.DeploymentCallbackFlow.prepare(DeploymentCallbackFlow.java:75)
	at weblogic.application.internal.flow.DeploymentCallbackFlow.prepare(DeploymentCallbackFlow.java:55)
	at weblogic.application.internal.BaseDeployment$1.next(BaseDeployment.java:729)
	at weblogic.application.utils.StateMachineDriver.nextState(StateMachineDriver.java:45)
	at weblogic.application.internal.BaseDeployment.prepare(BaseDeployment.java:241)
	at weblogic.application.internal.EarDeployment.prepare(EarDeployment.java:66)
	at weblogic.application.internal.DeploymentStateChecker.prepare(DeploymentStateChecker.java:158)
	at weblogic.deploy.internal.targetserver.AppContainerInvoker.prepare(AppContainerInvoker.java:65)
	at weblogic.deploy.internal.targetserver.AppDeployment.prepare(AppDeployment.java:166)
	at weblogic.management.deploy.internal.DeploymentAdapter$1.doPrepare(DeploymentAdapter.java:41)
	at weblogic.management.deploy.internal.DeploymentAdapter.prepare(DeploymentAdapter.java:193)
	at weblogic.management.deploy.internal.AppTransition$1.transitionApp(AppTransition.java:31)
	at weblogic.management.deploy.internal.ConfiguredDeployments$2.doItem(ConfiguredDeployments.java:741)
	at weblogic.management.deploy.internal.parallel.BucketInvoker$2.run(BucketInvoker.java:95)
	at weblogic.work.ContextWrap.run(ContextWrap.java:46)
	at weblogic.work.SelfTuningWorkManagerImpl$WorkAdapterImpl.run(SelfTuningWorkManagerImpl.java:670)
	at weblogic.invocation.ComponentInvocationContextManager._runAs(ComponentInvocationContextManager.java:352)
	at weblogic.invocation.ComponentInvocationContextManager.runAs(ComponentInvocationContextManager.java:337)
	at weblogic.work.LivePartitionUtility.doRunWorkUnderContext(LivePartitionUtility.java:57)
	at weblogic.work.PartitionUtility.runWorkUnderContext(PartitionUtility.java:41)
	at weblogic.work.SelfTuningWorkManagerImpl.runWorkUnderContext(SelfTuningWorkManagerImpl.java:644)
	at weblogic.work.ExecuteThread.execute(ExecuteThread.java:415)
	at weblogic.work.ExecuteThread.run(ExecuteThread.java:355)
Caused By: java.lang.SecurityException: class "com.microsoft.applicationinsights.web.javaee.RequestNameInterceptor_cs27u3_Impl"'s signer information does not match signer information of other classes in the same package
	at java.lang.ClassLoader.checkCerts(ClassLoader.java:898)
	at java.lang.ClassLoader.preDefineClass(ClassLoader.java:668)
	at java.lang.ClassLoader.defineClass(ClassLoader.java:761)
	at java.security.SecureClassLoader.defineClass(SecureClassLoader.java:142)
	at weblogic.utils.classloaders.GenericClassLoader.defineClassInternal(GenericClassLoader.java:1113)
	at weblogic.utils.classloaders.GenericClassLoader.defineClass(GenericClassLoader.java:1046)
	at weblogic.utils.classloaders.GenericClassLoader.findLocalClass(GenericClassLoader.java:1038)
	at weblogic.utils.classloaders.GenericClassLoader.findClass(GenericClassLoader.java:990)
	at weblogic.utils.classloaders.GenericClassLoader.doFindClass(GenericClassLoader.java:611)
	at weblogic.utils.classloaders.GenericClassLoader.loadClass(GenericClassLoader.java:543)
	at weblogic.utils.classloaders.GenericClassLoader.loadClass(GenericClassLoader.java:496)
	at weblogic.utils.classloaders.GenericClassLoader.loadClass(GenericClassLoader.java:473)
	at java.lang.Class.forName0(Native Method)
	at java.lang.Class.forName(Class.java:348)
	at weblogic.utils.classloaders.GenericClassLoader.defineCodeGenClass(GenericClassLoader.java:1360)
	at weblogic.ejb.container.ejbc.bytecodegen.GeneratorFactory.generateClass(GeneratorFactory.java:100)
	at weblogic.ejb.container.ejbc.bytecodegen.GeneratorFactory.generate(GeneratorFactory.java:44)
	at weblogic.ejb.container.ejbc.EJBCompiler.generate(EJBCompiler.java:577)
	at weblogic.ejb.container.ejbc.EJBCompiler.doCompile(EJBCompiler.java:186)
	at weblogic.ejb.container.ejbc.EJBCompiler.compileEJB(EJBCompiler.java:336)
	at weblogic.ejb.container.ejbc.EJBCompiler.compileEJB(EJBCompiler.java:304)
	at weblogic.ejb.container.deployer.EJBDeployer.compileEjbs(EJBDeployer.java:424)
	at weblogic.ejb.container.deployer.EJBDeployer.compileIfNecessary(EJBDeployer.java:355)
	at weblogic.ejb.container.deployer.EJBDeployer.prepare(EJBDeployer.java:723)
	at weblogic.ejb.container.deployer.EJBModule.prepare(EJBModule.java:405)
	at weblogic.application.internal.ExtensibleModuleWrapper$PrepareStateChange.next(ExtensibleModuleWrapper.java:295)
	at weblogic.application.internal.ExtensibleModuleWrapper$PrepareStateChange.next(ExtensibleModuleWrapper.java:285)
	at weblogic.application.utils.StateMachineDriver.nextState(StateMachineDriver.java:45)
	at weblogic.application.internal.ExtensibleModuleWrapper.prepare(ExtensibleModuleWrapper.java:109)
	at weblogic.application.internal.flow.ModuleListenerInvoker.prepare(ModuleListenerInvoker.java:100)
	at weblogic.application.internal.flow.ModuleStateDriver$1.next(ModuleStateDriver.java:192)
	at weblogic.application.internal.flow.ModuleStateDriver$1.next(ModuleStateDriver.java:187)
	at weblogic.application.utils.StateMachineDriver$ParallelChange.run(StateMachineDriver.java:83)
	at weblogic.application.utils.StateMachineDriver.nextStateInParallel(StateMachineDriver.java:144)
	at weblogic.application.internal.flow.ModuleStateDriver.parallelPrepare(ModuleStateDriver.java:46)
	at weblogic.application.internal.flow.DeploymentCallbackFlow.prepare(DeploymentCallbackFlow.java:75)
	at weblogic.application.internal.flow.DeploymentCallbackFlow.prepare(DeploymentCallbackFlow.java:55)
	at weblogic.application.internal.BaseDeployment$1.next(BaseDeployment.java:729)
	at weblogic.application.utils.StateMachineDriver.nextState(StateMachineDriver.java:45)
	at weblogic.application.internal.BaseDeployment.prepare(BaseDeployment.java:241)
	at weblogic.application.internal.EarDeployment.prepare(EarDeployment.java:66)
	at weblogic.application.internal.DeploymentStateChecker.prepare(DeploymentStateChecker.java:158)
	at weblogic.deploy.internal.targetserver.AppContainerInvoker.prepare(AppContainerInvoker.java:65)
	at weblogic.deploy.internal.targetserver.AppDeployment.prepare(AppDeployment.java:166)
	at weblogic.management.deploy.internal.DeploymentAdapter$1.doPrepare(DeploymentAdapter.java:41)
	at weblogic.management.deploy.internal.DeploymentAdapter.prepare(DeploymentAdapter.java:193)
	at weblogic.management.deploy.internal.AppTransition$1.transitionApp(AppTransition.java:31)
	at weblogic.management.deploy.internal.ConfiguredDeployments$2.doItem(ConfiguredDeployments.java:741)
	at weblogic.management.deploy.internal.parallel.BucketInvoker$2.run(BucketInvoker.java:95)
	at weblogic.work.ContextWrap.run(ContextWrap.java:46)
	at weblogic.work.SelfTuningWorkManagerImpl$WorkAdapterImpl.run(SelfTuningWorkManagerImpl.java:670)
	at weblogic.invocation.ComponentInvocationContextManager._runAs(ComponentInvocationContextManager.java:352)
	at weblogic.invocation.ComponentInvocationContextManager.runAs(ComponentInvocationContextManager.java:337)
	at weblogic.work.LivePartitionUtility.doRunWorkUnderContext(LivePartitionUtility.java:57)
	at weblogic.work.PartitionUtility.runWorkUnderContext(PartitionUtility.java:41)
	at weblogic.work.SelfTuningWorkManagerImpl.runWorkUnderContext(SelfTuningWorkManagerImpl.java:644)
	at weblogic.work.ExecuteThread.execute(ExecuteThread.java:415)
	at weblogic.work.ExecuteThread.run(ExecuteThread.java:355)
```

Any suggestions or advice would be greatly appreciated. 😄

**Comment by littleaj** [comment]
@ijmac3, I wasn't able to locate the exact issue, but I found a lead:
* This issue is with a different product, but the top of the stack trace is the same: https://www.oracle.com/webfolder/community/siebel/3605672.html
However, I can't seem to gain access to the page to view the solution 😞 . Perhaps you will have better luck.

I'll continue to look for leads, but I'm not very familiar with the deployment operation on WebLogic.
Here are some things I would investigate:

* It appears WebLogic is modifying the RequestNameInterceptor class and then attempting to validate against the original JAR signature. I would also try to find out why weblogic is modifying this class in the first place.
  * Perhaps this is not a necessary operation and can be disabled.
  * I think this hasn't happened previously because RequestNameInterceptor has the `@javax.interceptor.RequestNameInterceptor` annotation and could be loaded differently than other classes.
  * I would expect the same error with applicationinsights-web-<version>.jar. That one requires manual configuration of the filter, but also includes this annotated interceptor (-web-auto includes everything in -web plus some additional classes)
* Do interceptors require additional configuration, e.g. explicit declaration in a configuration/deplyment file?
* Perhaps a different JAR validation algorithm is needed for these generated classes. This could be configuration or the inclusion of a specific library/module.

I would also try to enable more verbose logging for this loading operation. Maybe it's actually attempting to validate against the wrong jar file.

I'll let you know if I find any more information. Please post any additional findings you think may be relevant.

**Comment by trask** [comment]
This should be resolved in [3.0 Preview](https://docs.microsoft.com/en-us/azure/azure-monitor/app/java-in-process-agent)
